### PR TITLE
refactor: improve channel edit typing

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

Refactors `GuildChannel._edit` to take kwargs instead of an `options` dict, and fixes a few minor typing issues found by refactoring. These changes are mostly internal, and shouldn't affect the public interface by much.

**Primary changes:**
- removed `_Undefined`, replaced usages with `utils.MISSING`
- refactored `GuildChannel._edit`, also simplifying parts of the code in the process
- replaced old channel edit method overloads with proper ones, with the return type depending on the provided parameters

**Other stuff:**
- removed `topic` parameter from `StageChannel.edit`, added `bitrate`
- added `**kwargs: NoReturn` to edit methods which preserves original runtime behavior for the most part but should help type-checking
    - this means you can (at runtime) use `<TextChannel>.edit(bitrate=...)` but not `<TextChannel>.edit(xyz=...)`; the former will do nothing, the latter will raise an exception, but type-checkers should show an error for both
- fixed `sync_permissions` parameter type of `VoiceChannel.edit` and `StageChannel.edit`
- made `slowmode_delay` and `default_auto_archive_duration` parameters optional
- changed `category` parameter type to allow any `Snowflake`-ish type instead of just `CategoryChannel`, no need for the unnecessary restriction as we only need the ID

## Checklist

- [ ] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [x] This PR fixes an issue
- [ ] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by onerandomusername** [comment]
Hi @shiftinv, what's the status on this PR?

**Comment by shiftinv** [comment]
> Hi @shiftinv, what's the status on this PR?

It was way behind master, rebased it for now but still missing both manual and possibly automated testing

File: disnake/abc.py [hunk 1]
```diff
@@ -94,21 +95,14 @@
...
 
     MessageableChannel = Union[GuildMessageable, DMChannel, PartialMessageable]
     SnowflakeTime = Union["Snowflake", datetime]
 
 MISSING = utils.MISSING
 
 
-class _Undefined:
-    def __repr__(self) -> str:
-        return "see-below"
```
Comment (onerandomusername) [inline disnake/abc.py:106]: I believe this was like this for sphinx documentation.

**Comment by shiftinv** [comment]
This should ~~finally~~ be ready now, ironed out the last few issues and improved the overloads.

**Review by EQUENOS — APPROVED** [review]

File: disnake/abc.py (hunk 1 above)
Comment (onerandomusername) [reply disnake/abc.py:108]: not really relevant.

**Pull request merged**